File: usaco/section_4.4/shuttle.py
"""
ID: huangch7
LANG: PYTHON3
TASK: shuttle
"""


task_name = "shuttle"
INF = 100000000


def get_ints_from_line():
    return [int(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()


fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

N = get_ints_from_line()[0]

# C(7, 3) * 4
# 7!/(3!*4!) * 4 = 7 * 6 * 5 * 4 = 840

start = [1] * N + [-1] + [0] * N
#target = tuple([0] * N + [-1] + [1] * N)



# target: (0, 0, 0, ..., _, 1, 1, 1...)
# mid  : (_, 1, 0, 1, 0, 1, 0, ... )
# mid. : (1, 0, 1, 0, ...,        ,_)
# start:(1, 1, 1, ..., _, 0, 0, 0...)

def display(alist):
    line = ""
    for x in alist:
        x = str(x) if x != -1 else "_"
        line += x
    # print(line)

def swap(i, j, alist):
    tmp = alist[i]
    alist[i] = alist[j]
    alist[j] = tmp
    display(alist)

display(start)
ans = [N]
swap(N - 1, N, start)
# display(start)
# 搬运1个0
state = [0, 1]
p = N - 1
while state[1] < N:
    n = state[1]
    while n:
        nextp = p + 2 if state[0] == 0 else p - 2
        ans.append(nextp + 1)
        swap(p, nextp, start)
        p = nextp
        n -= 1
    nextp = p + 1 if state[0] == 0 else p - 1
    ans.append(nextp + 1)
    swap(p, nextp, start)
    p = nextp

    state = [1 - state[0], state[1] + 1]

# N = 4
# 4 6 7 5 3 2 4 6 8 9 7 5 3 1 2 4 6 8 7 5
# 3 4 6 5
# N = 6
# 6 8 9 7 5 4 6 8 10 11 9 7 5 3 2 4 6 8 10 12
# 13 11 9 7 5 3 1 2 4 6 8 10 12 11 9 7 5 3 4 6
# 8 10 9 7 5 6 8 7
#print(state)

while state[1] > 0:
    n = state[1]
    while n:
        nextp = p + 2 if state[0] == 0 else p - 2
        ans.append(nextp + 1)
        swap(p, nextp, start)
        p = nextp
        n -= 1
    nextp = p - 1 if state[0] == 0 else p + 1
    ans.append(nextp + 1)
    swap(p, nextp, start)
    p = nextp

    state = [1 - state[0], state[1] - 1]
    
    

# print(ans)




final_ans = ans

for i in range(0, len(final_ans), 20):
    fout.write(" ".join([str(x) for x in final_ans[i: i + 20]]))
    fout.write("\n")

        
    

fout.close()

File: usaco/training/section_4.4/shuttle.cpp
/*
ID: huangch7
LANG: C++
TASK: shuttle
*/  
#include <iostream>
#include <fstream>
#include <queue>
#include <unordered_map>
#include <array>

using namespace std;

const int INF = 100000000;

unordered_map<string, int> state2id;
unordered_map<int, string> id2state;

int get_idx(string x){
  if (state2id.find(x) == state2id.end()){
    id2state[state2id.size()] = x;
    state2id[x] = state2id.size();
  }
  return state2id[x];
}

string get_state(int x){
  return id2state[x];
}

int main() {
    ifstream fin("shuttle.in");
    ofstream fout("shuttle.out");

    int N;
    fin >> N;


    string target = "";
    for (int i = 0; i < N; ++i) target += "1";
    target += "A";
    for (int i = 0; i < N; ++i) target += "0";

    string start = "";
    for (int i = 0; i < N; ++i) start += "0";
    start += "A";
    for (int i = 0; i < N; ++i) start += "1";

    unordered_map<int, int> next;
    unordered_map<string, int> ans;

    queue<int> q;
    q.push(get_idx(target));
    next[get_idx(target)] = -1;
    ans[target] = -1;

    while (!q.empty()) {
      string cur = get_state(q.front());
        q.pop();

        int black = -1;
        for (int i = 0; i < cur.size(); ++i) {
            if (cur[i] == 'A') {
                black = i;
                break;
            }
        }

        for (int i : {-2, -1, 1, 2}) {
            int idx = black + i;
            if (idx < 0 || idx >= 2 * N + 1) {
                continue;
            }

            string state = cur;
            swap(state[black], state[idx]);

            if (next.find(get_idx(state)) != next.end()) {
                continue;
            }

            next[get_idx(state)] = get_idx(cur);
            ans[state] = idx;

            if (state == start) {
                break;
            }

            q.push(get_idx(state));
        }
    }

    vector<int> final_ans_;
    string cur = start;
    while (cur != target) {
        final_ans_.push_back(ans[cur] + 1);
        cur = get_state(next[get_idx(cur)]);
    }
    
    vector<int> final_ans;
    for (int i = final_ans_.size() - 1; i >= 0; --i) {
      //      cout << i << " " << final_ans_[i] << endl;
      final_ans.push_back(final_ans_[i]);
    }
    
    for (size_t i = 0; i < final_ans.size(); i += 20) {
        for (size_t j = i; j < min(i + 20, final_ans.size()); ++j) {
          if (j != i) fout << " ";
          fout << final_ans[j];
        }
        fout << "\n";
    }

    fout.close();
    fin.close();

    return 0;
}
